### ISSUE TITLE
Empty or NaN data pollute gradient even if they are not involved during backward

### ISSUE THREAD

**Issue by prclibo** [pytorch/pytorch#70342]
### 🐛 Describe the bug

Hi, Probably this is not a bug, but I am just wondering how the behavior is caused and if it could be improved.

Say I have 2 pieces of data in a batch. One is valid and the other is NaN. I pass it to my network and get a batch of two losses. Obviously, one loss is valid and the other is NaN. I backward only the valid loss and expect the training to still work good, but it turns out that the training is polluted by NaN even if I did not backward the NaN loss. Please see the following code piece:

```
import numpy as np
import torch
from torch import nn

embed_dim = 512
batch_size = 2
token_num = 20
iter_num = 5

class FC(nn.Module):
    def __init__(self):
        super(FC, self).__init__()
        self.fc = nn.Linear(embed_dim, 1)

    def forward(self, x):
        return self.fc(x).flatten(1, -1).sum(1)

def train(data, model, optimizer):
    for i in range(iter_num):
        optimizer.zero_grad()
        l = model(data[i])
        print(i, l)

        l[0].backward() # WE ONLY BACKWARD THE LOSS CORRESPONDING TO THE VALID DATA
        optimizer.step()

data = torch.rand(iter_num, batch_size, embed_dim)

seed = 8; np.random.seed(seed); torch.manual_seed(seed); torch.cuda.manual_seed(seed)
model = FC()
optimizer = torch.optim.AdamW(model.parameters())
train(data, model, optimizer)
print('----')

seed = 8; np.random.seed(seed); torch.manual_seed(seed); torch.cuda.manual_seed(seed)
model = FC()
optimizer = torch.optim.AdamW(model.parameters())
data[:, 1:] = np.nan # Set an invalid piece of NaN data
train(data, model, optimizer)
```

I got output like: 
```
0 tensor([-0.2391, -0.3819], grad_fn=<SumBackward1>)
1 tensor([-0.5708, -0.7382], grad_fn=<SumBackward1>)
2 tensor([-0.5046, -0.7748], grad_fn=<SumBackward1>)
3 tensor([-0.9986, -0.9386], grad_fn=<SumBackward1>)
4 tensor([-1.2657, -1.0637], grad_fn=<SumBackward1>)
----
0 tensor([-0.2391,     nan], grad_fn=<SumBackward1>)
1 tensor([nan, nan], grad_fn=<SumBackward1>)
2 tensor([nan, nan], grad_fn=<SumBackward1>)
3 tensor([nan, nan], grad_fn=<SumBackward1>)
4 tensor([nan, nan], grad_fn=<SumBackward1>)
```

For the second case, the model weight is polluted as NaN from the #1 iter. Similar behavior is also observed for partially empty sequence in MultiHeadAttention module.

### Versions

PyTorch 1.8

**Comment by Lezcano** [comment]
Alas, this is a known limitation of the autograd system in PyTorch. See https://github.com/pytorch/pytorch/issues/68425

**Comment by albanD** [comment]
This is expected I am afraid.
I would encourage you to check the masked Tensor prototype mentioned in the linked issue if this is problematic for you.

Also if you have a specific use case where you want help figuring out a workaround, feel free to ask about it on the forum.
